pkg/volume/glusterfs/glusterfs.go

```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	GlusterRestvolpath string `json:"path"`
+	GlusterRestUrl     string `json:"resturl"`
+	GlusterRestAuth    bool   `json:"restauthenabled"`
+	GlusterRestUser    string `json:"restuser"`
+	GlusterRestUserKey string `json:"restuserkey"`
+}
+
+type glusterfsVolumeProvisioner struct {
+	*glusterfsMounter
+	*glusterfsClusterConf
+	options volume.VolumeOptions
+}
+
+var clusterconf = new(glusterfsClusterConf)
+
+func (plugin *glusterfsPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec)
+}
+
+func (plugin *glusterfsPlugin) newDeleterInternal(spec *volume.Spec) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.Glusterfs == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.Spec.Glusterfs is nil")
+	}
+	return &glusterfsVolumeDeleter{
+		glusterfsMounter: &glusterfsMounter{
+			glusterfs: &glusterfs{
+
+				volName: spec.Name(),
+				plugin:  plugin,
+			},
+
+			path: spec.PersistentVolume.Spec.Glusterfs.Path,
+		}}, nil
+}
+
+type glusterfsVolumeDeleter struct {
+	*glusterfsMounter
+	*glusterfsClusterConf
+}
+
+func (d *glusterfsVolumeDeleter) GetPath() string {
+	name := glusterfsPluginName
+	return d.plugin.host.GetPodVolumeDir(d.glusterfsMounter.glusterfs.pod.UID, strings.EscapeQualifiedNameForDisk(name), d.glusterfsMounter.glusterfs.volName)
+}
+
+func (d *glusterfsVolumeDeleter) Delete() error {
+	return d.DeleteVolume()
+}
+
+func (d *glusterfsVolumeDeleter) DeleteVolume() error {
+	glog.V(1).Infof("glusterfs: delete volume :%s ", d.glusterfsMounter.path)
+	volumetodel := d.glusterfsMounter.path
+	d.glusterfsClusterConf = clusterconf
+	newvolumetodel := dstrings.TrimPrefix(volumetodel, "vol_")
+	cli := gcli.NewClient(d.GlusterRestUrl, d.GlusterRestUser, d.GlusterRestUserKey)
+	if cli == nil {
+		glog.V(1).Infof("glusterfs: failed to create gluster rest client")
+		return fmt.Errorf("glusterfs: failed to create gluster rest client")
+	}
+	err := cli.VolumeDelete(newvolumetodel)
+	if err != nil {
+		glog.V(1).Infof("glusterfs: error when deleting the volume :%s", err)
+		return err
+	}
+	glog.V(1).Infof("glusterfs: volume %s deleted successfully", volumetodel)
+	return nil
+
+}
+
+func (r *glusterfsVolumeProvisioner) Provision(pv *api.PersistentVolume) error {
+
+	config := r.glusterfsMounter.plugin.host.GetStorageConfigDir()
+	if config == "" {
+		glog.V(1).Infof("glusterfs: no cluster storage config file provided")
+		return fmt.Errorf("glusterfs: no gluster cluster storage config provided")
+	}
+	file := path.Join(config, "gluster.json")
+	fp, err := os.Open(file)
+	if err != nil {
+		glog.V(1).Infof("glusterfs cluster configuration file open err %s/%s", file, err)
+		return fmt.Errorf("glusterfs cluster configuration file open err %s/%s", file, err)
+	}
+	defer fp.Close()
+
+	decoder := json.NewDecoder(fp)
+	if err = decoder.Decode(clusterconf); err != nil {
+		glog.V(1).Infof("glusterfs: decode err: %s.", err)
```

Comment:
`glusterfs: cannot parse file %s: %s`. It's important to show the file name, otherwise the user it lost.